src/Swarm/Game/Step.hs

```diff
@@ -913,12 +919,52 @@ execConst c vs s k = do
         -- set the copied robot to be a "system" robot so it is
         -- capable of executing any commands; the As command
         -- already requires "God" capability.
-        v <-
-          evalState @Robot (r & systemRobot .~ True) . evalState @GameState g $
-            runCESK (Out prog s [FApp (VCApp Force []), FExec])
+
+        -- -----------------------------------------------------------------------------
+        -- TODO: move this code out
+        -- Slightly edited copy of Unit.hs version of runCESK:
+        let runCESK' :: CESK -> StateT Robot (StateT GameState IO) (Either Text Value)
+            runCESK' (Up exn _ []) = return (Left (formatExn exn))
+            runCESK' cesk = case finalValue cesk of
+              Just (v, _) -> return (Right v)
+              Nothing -> do
+                cesk1 <- stepCESK cesk
+                ticks += 1
+                runCESK' cesk1
+        -- -----------------------------------------------------------------------------
+        let evalAs :: IO (Either Text Value)
+            evalAs =
+              Out prog s [FApp (VCApp Force []), FExec]
+                & runCESK'
+                & flip evalStateT (r & systemRobot .~ True)
+                & flip evalStateT (g & gameMode .~ CreativeMode)
+
+        va <-
+          sendIO . Async.async $
+            either VString id -- TODO: that is not right...
+              <$> evalAs
 
         -- Return the value returned by the hypothetical command.
-        return $ Out v s k
+        i <- gensym <+= 1
+        return $ Out (VAsync $ GAsync i va) s (FApp (VCApp Poll []) : k)
```

Comment:
Conceptually, it probably makes more sense to make a new kind of `Frame` called `FPoll` rather than making `Poll` a primitive function.

Reply:
I went the other way and made `poll` a real command. :man_shrugging: